Improve unstocking inventory units from stock locations

AlessioRocco opened the pull request:
**Description**

This PR fixes the bug reported from this issue https://github.com/solidusio/solidus/issues/3527.

The `Spree::Stock::InventoryUnitsFinalizer#unstock_inventory_units` method was unstocking the `line_item` original `variant` but not the `variant` set in the `inventory_unit`, this PR fixes this error and improve the code a bit.

**Checklist:**
- [x] I have followed [Pull Request guidelines](https://github.com/solidusio/solidus/blob/master/CONTRIBUTING.md#pull-request-guidelines)
- [x] I have added a detailed description into each commit message

kennyadsl commented:
@AlessioRocco do we have any performance improvements with this refactor as well?

AlessioRocco commented:
@kennyadsl I thought this PR shouldn't be related to the performance, but the fact that the `line item` is used to retrieve the `variant` when we have a direct association on the `inventory_unit` makes confusion to me. 

With this PR I'm trying to simplify the code and reduce this confusion.

It also allows creating a shipment with products that differ from the variant in the line item.

Suppose that you have a variant that represents a pack of products, with the old implementation you can't adapt the code to create a shipment that sends something different from the line item variant instead you can do it with the code proposed in this PR.

kennyadsl commented:
Before merging, maybe we should add a test to describe this behavior as well?

> Suppose that you have a variant that represents a pack of products, with the old implementation you can't adapt the code to create a shipment that sends something different from the line item variant instead you can do it with the code proposed in this PR.

sechix commented:
Hi, Could we merge this commit @AlessioRocco , change size is not working ok and I need it . Thanks!

kennyadsl commented:
@sechix I look forward to merging this but I'd like to wait for some specs for this behavior. This way we'll be sure to don't reintroduce the bug in future changes.

kennyadsl commented:
@AlessioRocco any news here?

AlessioRocco commented:
@kennyadsl I've added the specs and squashed the commits into one to make the change clear for the future.
I've also changed the PR title and description, can you please link the related issue https://github.com/solidusio/solidus/issues/3527 to this PR so it will be closed when merge?
